File: 1185-find-in-mountain-array/find-in-mountain-array.py
# """
# This is MountainArray's API interface.
# You should not implement it, or speculate about its implementation
# """
#class MountainArray:
#    def get(self, index: int) -> int:
#    def length(self) -> int:

class Solution:
    def findInMountainArray(self, target: int, mountainArr: 'MountainArray') -> int:

        # FIND PEAK 

        l = 1 
        r = mountainArr.length() - 2

        while l <= r:

            m = l + (r - l) // 2

            left = mountainArr.get( m - 1 )
            mid = mountainArr.get( m )
            right = mountainArr.get( m + 1 )

            if left < mid < right:
                l = m + 1
            elif left > mid > right:
                r = m - 1
            else:
                break
        
        # LEFT PORTION 

        peak = m 

        l = 0 
        r = peak - 1 

        while l <= r:
            m = l + ( r - l ) // 2

            val = mountainArr.get(m)

            if val > target:
                r = m - 1
            elif val < target:
                l = m + 1
            else:
                return m 
        
        #RIGHT PORTION 

        l = peak 
        r = mountainArr.length() -1 

        while l <= r:
            m =  l + (r - l) // 2

            val = mountainArr.get(m)

            if val > target:
                l = m + 1
            elif val < target:
                r = m - 1
            else:
                return m 
        return -1 
        


        

File: 1185-find-in-mountain-array/find-in-mountain-array.java
class Solution {

    public int findInMountainArray(int target, MountainArray mountainArr) {

        int l = 1;
        int r = mountainArr.length() - 2;

        int peak = -1;

        while (l <= r) {

            int m = l + (r - l) / 2;

            int left = mountainArr.get(m - 1);
            int mid = mountainArr.get(m);
            int right = mountainArr.get(m + 1);

            if (left < mid && mid < right) {

                l = m + 1;

            } else if (left > mid && mid > right) {

                r = m - 1;

            } else {

                peak = m;
                break;
            }
        }

        l = 0;
        r = peak;

        while (l <= r) {

            int m = l + (r - l) / 2;

            int val = mountainArr.get(m);

            if (val > target) {

                r = m - 1;

            } else if (val < target) {

                l = m + 1;

            } else {

                return m;
            }
        }

        l = peak + 1;
        r = mountainArr.length() - 1;

        while (l <= r) {

            int m = l + (r - l) / 2;

            int val = mountainArr.get(m);

            if (val > target) {

                l = m + 1;

            } else if (val < target) {

                r = m - 1;

            } else {

                return m;
            }
        }

        return -1;
    }
}